pkg/block/block.go

```diff
@@ -95,3 +101,59 @@ func renameFile(from, to string) error {
 	}
 	return pdir.Close()
 }
+
+// DownloadDir downloads directory that is mean to be block directory (ends with ULID and contains block data).
+func DownloadDir(ctx context.Context, bucket objstore.Bucket, blockDir, dst string) error {
+	if _, err := ulid.Parse(path.Base(blockDir)); err != nil {
+		return errors.Errorf("given directory is not block directory %s", blockDir)
+	}
+
+	if err := objstore.DownloadDir(ctx, bucket, blockDir, dst); err != nil {
+		return err
+	}
+
+	chunksDir := filepath.Join(dst, "chunks")
+	_, err := os.Stat(chunksDir)
+	if os.IsNotExist(err) {
+		// This can happen if block is empty. We cannot easily upload empty directory, so create one here.
+		return os.Mkdir(chunksDir, os.ModePerm)
+	}
+
+	if err != nil {
+		return errors.Wrapf(err, "stat %s", chunksDir)
+	}
+
+	return nil
+}
+
+// DownloadMeta downloads only meta file from bucket by block ID.
+func DownloadMeta(ctx context.Context, bkt objstore.Bucket, id ulid.ULID) (Meta, error) {
+	rc, err := bkt.Get(ctx, path.Join(id.String(), "meta.json"))
+	if err != nil {
+		return Meta{}, errors.Wrapf(err, "meta.json bkt get for %s", id.String())
+	}
+	defer rc.Close()
+
+	// Do a full decode/encode cycle to ensure we only print valid JSON.
+	var m Meta
+
+	if err := json.NewDecoder(rc).Decode(&m); err != nil {
+		return Meta{}, errors.Wrapf(err, "decode meta.json for block %s", id.String())
+	}
+	return m, nil
+}
+
+// Foreach runs doFn for each block ID in the given bucket.
+func Foreach(ctx context.Context, bucket objstore.Bucket, doFn func(ulid.ULID) error) error {
```

Comment:
I do find it a bit confusing that the `block` package now has a mix of functions that work against an object storage bucket or the file system. While having no naming pattern telling us which is which.

I think we've at least two instances of helper functions that do the exact some thing as this one just against a local dir.
So the question is why this helper gets to be a first-class function while the other is duplicated as needed.

Reply:
Don't follow ):

Reply:
My point is that we have equivalent functionality which we need for obj storage and the FS alike.
As you said out-of-band, we could address this with a FS implementation of a bucket. We just need to be careful that it doesn't become awkward to use.

In this particular case, I wonder whether we are better of simply providing a helper for the ULID validation, i.e. the 2 if conditions here and have callers keep calling `bkt.Iter` directly and using the helper in there.

Reply:
Still not too happy about that one. Maybe having a function here:

```go
func IsBlockDir(path string) bool {
    _, err := ulid.Parse(path.Base(path))
    return err == nil
}
```

and just calling `bucket.Iter` in our code as before is clean and simple:

```go
bucket.Iter(ctx, "", func(name string) error {
    if !block.IsBlockDir(name) {
        return nil
    }
    // ...
})